### PR TITLE
Control the start of event reporting for expiring certificates

### PR DESCRIPTION
**Pull request by allanrogerr**
### Issue summary

In cases where users install certificates of short duration, kubernetes events are spammed with multiple warnings of certificate expiry. This is undesirable. A means on reducing the number of warnings produced is more desirable.
See https://github.com/minio/operator/issues/1791

In the tenant crd, `.spec.certExpiryAlertThreshold` represents the minimum number of days to expiry before an alert for an expiring certificate is fired.
For example considering `.spec.certExpiryAlertThreshold = 1`, if a given certificate will expire in 7 days then expiration events will only be triggered 1 day before expiry.

### Test summary

1. Build and install the minio operator
2. Apply a tenant with Helm or Kustomize
3. Create expired and expiring certificates
4. Create secrets based on these
5. Apply the secrets as external certificates
6. Experiment with different values of `certExpiryAlertThreshold` e.g.
7. Observe

* if certExpiryAlertThreshold is not present then all certificate expiry events are reported. This is current functionality
* if the whole number of days left for a certificate to expire is less than certExpiryAlertThreshold then the certificate expiry events are reported
* if the whole number of days left for a certificate to expire is greater than or equal to certExpiryAlertThreshold then the certificate expiry events are not reported
* expired certificate events are always reported

Verbose tests here:
https://github.com/allanrogerr/public/wiki/operator%E2%80%901791

Fixes: https://github.com/minio/operator/issues/1791

### CODE REVIEW

**Review by pjuarezd — APPROVED** [review]

File: helm/operator/templates/minio.min.io_tenants.yaml [hunk 1]
```diff
@@ -875,6 +875,9 @@ spec:
                       type: string
                     type: array
                 type: object
+              certExpiryAlertThreshold:
+                format: int64
```
Comment (ramondeklein) [inline helm/operator/templates/minio.min.io_tenants.yaml:879]: Please use `int32` (it's in days, so `int64` is not needed). Marshalling 64-bit integers to Javascript is a pain, because it makes it a string. Using `int32` will keep it a number and keep things simple.

File: pkg/controller/custom.go [hunk 2]
```diff
@@ -108,17 +108,21 @@ func (c *Controller) getCustomCertificates(ctx context.Context, tenant *miniov2.
...
-				}
-				if expiresInDays > 0 && expiresInDays < 10 {
-					c.recorder.Event(tenant, corev1.EventTypeWarning, "CertificateExpiryImminent", fmt.Sprintf("%s certificate '%s' is expiring in %d days", certType, secret.Name, expiresInDays))
+				if tenant.Spec.CertExpiryAlertThreshold == nil || expiresInDays < *tenant.Spec.CertExpiryAlertThreshold {
+					if expiresInDays >= 10 && expiresInDays < 30 {
+						c.recorder.Event(tenant, corev1.EventTypeWarning, "CertificateExpiring", fmt.Sprintf("%s certificate '%s' is expiring in %d days", certType, secret.Name, expiresInDays))
+					}
+					if expiresInDays > 0 && expiresInDays < 10 {
+						c.recorder.Event(tenant, corev1.EventTypeWarning, "CertificateExpiryImminent", fmt.Sprintf("%s certificate '%s' is expiring in %d days", certType, secret.Name, expiresInDays))
+					}
```
Comment (ramondeklein) [inline pkg/controller/custom.go:117]: Shouldn't these values be related to the `Spec.CertExpiryAlertThreshold` value?

File: resources/base/crds/minio.min.io_tenants.yaml [hunk 3]
```diff
@@ -875,6 +875,9 @@ spec:
                       type: string
                     type: array
                 type: object
+              certExpiryAlertThreshold:
+                format: int64
```
Comment (ramondeklein) [inline resources/base/crds/minio.min.io_tenants.yaml:879]: Same here.

**Review by ramondeklein — CHANGES_REQUESTED** [review]

File: helm/operator/templates/minio.min.io_tenants.yaml (hunk 1 above)
Comment (allanrogerr) [reply helm/operator/templates/minio.min.io_tenants.yaml:879]: I kept this as `int64` (1) in order to simplify the comparison between `certExpiryAlertThreshold` and `int64` var `expiresInDays`. Also, a separate helper function (2) would be needed to convert the possibly null (3)`*tenant.Spec.CertExpiryAlertThreshold` to `int64`, slowing down the program execution.
See https://github.com/minio/operator/blob/0b2bacb04bc0329eb1d16bfdfe7b8dd4ac1caa80/pkg/controller/custom.go#L111

In summary, using `int64` simplifies reading the code, reduces the amount of code needed, and does not affect the speed of execution as much.

Comment (allanrogerr) [reply helm/operator/templates/minio.min.io_tenants.yaml:879]: @ramondeklein Your thoughts on these reasons versus the marshalling complexity?

**Review by shtripat — APPROVED** [review]
lgtm, verified

File: helm/operator/templates/minio.min.io_tenants.yaml (hunk 1 above)
Comment (ramondeklein) [reply helm/operator/templates/minio.min.io_tenants.yaml]: @allanrogerr Good to have it as `int32`. I'll elaborate why I prefer this...

The problem with the default marshalling is that it sends `int64` protobuf via JSON as a `string`. Javascript handles each number as a 64-bit float (IEEE-754), where it uses 11 bits for the exponent. Because of that it cannot handle 64-bit integers with the proper precision and therefor marshalling reverts to using `string`. This requires both the Go and Javascript code to convert back/forth numbers. That's why I prefer using `int32` if it fits.

If it only saves a cast in Go from `int32` to `int64` (that will never loose precision), then I would prefer that instead. Converting from a string to a number always requires error-checking and dealing with the error. So if the semantics allow for `int32`, then I would use that instead.

File: pkg/controller/custom.go (hunk 2 above)
Comment (ramondeklein) [reply pkg/controller/custom.go:117]: Some companies have rather short certificate expirations and rotate certificates often, so that's probably why this PR is needed. I thought we wanted to control when these messages occur, but I guess this is fine too. Maybe @jiuker is right and another name would make it more obvious.

File: pkg/controller/custom.go [hunk 4]
```diff
@@ -102,23 +102,27 @@ func (c *Controller) getCustomCertificates(ctx context.Context, tenant *miniov2.
...
+					if expiresIn <= 0 {
+						c.recorder.Event(tenant, corev1.EventTypeWarning, "CertificateExpired", fmt.Sprintf("%s certificate '%s' has expired", certType, secret.Name))
+					}
```
Comment (ramondeklein) [inline pkg/controller/custom.go:120]: Although the `expiresInDays < *tenant.Spec.CertExpiryAlertThreshold` probably cannot be set to a negative number, I would prefer to move this out of the outer `if` statement. We should always report about expired certificates. No matter what the thresholds are. Maybe even raise the event type to error?

**Review by ramondeklein — CHANGES_REQUESTED** [review]
Would like to see the expiration check to be moved out of the outer `if` to ensure we're always reporting that. We have already spent a lot of Subnet hours, because of unnoticed certificate expirations.

File: pkg/controller/custom.go (hunk 4 above)
Comment (allanrogerr) [reply pkg/controller/custom.go:120]: `expiresInDays` is signed and can be negative. Expired certificates are always reported.

Changing the event type to error is not within the scope of this PR.

**Review by ramondeklein — APPROVED** [review]

**Review by shtripat — APPROVED** [review]
